lib/collection/src/collection/mmr.rs

```diff
@@ -0,0 +1,800 @@
+    if num_vectors < 2 {
+        debug_assert!(
+            false,
+            "There should be at least two vectors to calculate similarity matrix"
+        );
+        return Err(CollectionError::service_error(
+            "There should be at least two vectors to calculate similarity matrix",
+        ));
+    }
+
+    // Initialize similarity matrix with zeros
+    let mut similarity_matrix = vec![vec![0.0; num_vectors]; num_vectors];
+
+    // Prepare all scorers
+    let raw_scorers = vectors
+        .into_iter()
+        .map(|vector| {
+            let query = QueryVector::Nearest(vector);
+            new_raw_scorer(
+                query,
+                volatile_storage,
+                hw_measurement_acc.get_counter_cell(),
+            )
+        })
+        .collect::<OperationResult<Vec<_>>>()?;
+
+    // Compute similarities only for upper triangle to optimize (i < j)
+    // Since similarity is symmetric: sim(i,j) == sim(j,i), we can avoid duplicate computation
+    for i in 0..num_vectors {
+        // Only compute scores for the upper triangle
+        let upper_offsets: Vec<u32> = ((i + 1)..num_vectors).map(|j| j as u32).collect();
+
+        if !upper_offsets.is_empty() {
+            let mut scores = vec![0.0; upper_offsets.len()];
+            raw_scorers[i].score_points(&upper_offsets, &mut scores);
+
+            for (&vector_idx, similarity) in upper_offsets.iter().zip(scores) {
+                let j = vector_idx as usize;
+                // Set both (i,j) and (j,i) since similarity is symmetric
+                similarity_matrix[i][j] = similarity;
+                similarity_matrix[j][i] = similarity;
+            }
+        }
+        // Diagonal elements remain 0.0 (self-similarity excluded)
+    }
+
+    Ok(similarity_matrix)
+}
+
+/// Maximum Marginal Relevance (MMR) algorithm
+///
+/// Iteratively selects points by considering their similarity to
+/// already selected points, combining diversity and relevance.
+///
+/// # Arguments
+///
+/// * `candidates` - the list of points to select from
+/// * `query_similarities` - similarities to the query for each candidate. Offsets refer to the index of the candidate in the `candidates` vector.
+/// * `similarity_matrix` - full pairwise similarity matrix between candidates
+/// * `lambda` - the lambda parameter for the MMR algorithm (0.0 = max diversity, 1.0 = max relevance)
+/// * `limit` - the maximum number of points to select
+pub fn maximum_marginal_relevance(
+    candidates: Vec<ScoredPoint>,
+    query_similarities: Vec<ScoreType>,
+    similarity_matrix: &[Vec<ScoreType>],
+    lambda: f32,
+    limit: usize,
+) -> Vec<ScoredPoint> {
+    let num_candidates = candidates.len();
+    if num_candidates == 0 || limit == 0 {
+        return Vec::new();
+    }
+
+    let mut selected_indices = Vec::with_capacity(limit);
+    let mut remaining_indices: AHashSet<usize> = (0..num_candidates).collect();
```

Comment:
Alternatively we could use `IndexSet` here since it's better optimized for iterating - which we seem to do a lot.